java/org/apache/catalina/realm/JNDIRealm.java

```diff
@@ -1419,18 +1469,27 @@ protected User getUser(JNDIConnection connection, String username, String creden
         User user = null;
 
         // Get attributes to retrieve from user entry
-        List<String> list = new ArrayList<>();
-        if (userPassword != null) {
-            list.add(userPassword);
-        }
-        if (userRoleName != null) {
-            list.add(userRoleName);
-        }
-        if (userRoleAttribute != null) {
-            list.add(userRoleAttribute);
+        // Includes attributes required for authentication and authorization as well as
+        // user attributes to additionally query from the user's directory entry
+        String[] attrIds = null;
+        if (userAttributesList == null
+                || !userAttributesList.get(0).equals(USER_ATTRIBUTES_WILDCARD)) {
+            Set<String> list = new HashSet<>();
```

Comment:
A hash set is not a list. You should use consistent collections throughout your PR.

Reply:
Ah... just reused an existing variable and changed from List to Set (in order to remove duplicates). Forgot to rename that accordingly.